pkg/plugins/document/mongodb/mongodb.go

```diff
@@ -333,62 +330,119 @@ func (s *MongoDocService) Query(collection *document.Collection, expressions []d
+				"error decoding mongo document",
 				err,
 			)
 		}
 
-		id := docSnap[primaryKeyAttr].(string)
+		queryResult.Documents = append(queryResult.Documents, *sdkDoc)
 
-		// remove id from content
-		delete(docSnap, primaryKeyAttr)
+		// If query limit configured determine continue tokens
+		if limit > 0 && len(queryResult.Documents) == limit {
+			tokens := ""
+			if orderBy != "" {
+				tokens = fmt.Sprintf("%v", sdkDoc.Content[orderBy]) + "|"
+			}
+			tokens += sdkDoc.Key.Id
 
-		sdkDoc := document.Document{
-			Content: docSnap,
-			Key: &document.Key{
-				Collection: collection,
-				Id:         id,
-			},
+			queryResult.PagingToken = map[string]string{
+				"pagingTokens": tokens,
+			}
 		}
+	}
 
-		if docSnap[parentKeyAttr] != nil {
-			parentId := docSnap[parentKeyAttr].(string)
+	return queryResult, nil
+}
 
-			sdkDoc.Key.Collection = &document.Collection{
-				Name: collection.Name,
-				Parent: &document.Key{
-					Collection: collection.Parent.Collection,
-					Id:         parentId,
-				},
-			}
+func (s *MongoDocService) QueryStream(collection *document.Collection, expressions []document.QueryExpression, limit int) document.DocumentIterator {
+	newErr := errors.ErrorsWithScope(
+		"MongoDocService.QueryStream",
+		map[string]interface{}{
+			"collection": collection,
+		},
+	)
+
+	colErr := document.ValidateQueryCollection(collection)
+	expErr := document.ValidateExpressions(expressions)
 
-			delete(docSnap, parentKeyAttr)
+	if colErr != nil || expErr != nil {
+		// Return an error only iterator
+		return func() (*document.Document, error) {
+			return nil, newErr(
+				codes.InvalidArgument,
+				"invalid arguments",
+				fmt.Errorf("collection error:%v, expression error: %v", colErr, expErr),
+			)
 		}
+	}
 
-		queryResult.Documents = append(queryResult.Documents, sdkDoc)
+	cursor, _, cursorErr := s.getCursor(collection, expressions, limit, nil)
 
-		// If query limit configured determine continue tokens
-		if limit > 0 && len(queryResult.Documents) == limit {
-			tokens := ""
-			if orderByAttrib != "" {
-				tokens = fmt.Sprintf("%v", docSnap[orderByAttrib]) + "|"
-			}
-			tokens += id
+	return func() (*document.Document, error) {
+		if cursorErr != nil {
+			return nil, cursorErr
+		}
 
-			queryResult.PagingToken = map[string]string{
-				"pagingTokens": tokens,
+		if cursor.Next(s.context) {
+			// return the next document
+			return mongoDocToDocument(collection, cursor)
+		} else {
+			// there was an error
+			// Close the cursor
+			cursor.Close(s.context)
+
+			// Examine the cursors error to see if it's exhausted
+			if cursor.Err() != nil {
+				return nil, cursor.Err()
+			} else {
+				return nil, io.EOF
 			}
 		}
 	}
+}
 
-	return queryResult, nil
+func mongoDocToDocument(coll *document.Collection, cursor *mongo.Cursor) (*document.Document, error) {
+	var docSnap map[string]interface{}
+
+	err := cursor.Decode(&docSnap)
+
+	if err != nil {
+		return nil, fmt.Errorf("error decoding mongo document")
```

Comment:
I think we should just return the error here, to avoid eating error the root cause.

Reply:
done